File: backend/app/api/v1/workspaces.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user_id, get_current_workspace
from app.db.models.workspace import Workspace
from app.db.session import get_session
from app.schemas.workspace import WorkspaceCreate, WorkspacePlanUpdate, WorkspaceRead
from app.services import quota_service, workspace_service

router = APIRouter(prefix="/workspaces", tags=["workspaces"])


async def _to_read_model(session: AsyncSession, workspace: Workspace) -> WorkspaceRead:
    leads_used = await quota_service.count_leads(session, workspace.id)
    seats_used = await quota_service.count_seats(session, workspace.id)
    return WorkspaceRead(
        id=workspace.id,
        name=workspace.name,
        plan=workspace.plan,
        lead_quota=workspace.lead_quota,
        seat_quota=workspace.seat_quota,
        leads_used=leads_used,
        seats_used=seats_used,
        created_at=workspace.created_at
    )


@router.post("", response_model=WorkspaceRead, status_code=status.HTTP_201_CREATED)
async def create_workspace(
    payload: WorkspaceCreate,
    user_id: Annotated[str, Depends(get_current_user_id)],
    session: Annotated[AsyncSession, Depends(get_session)]
):
    workspace = await workspace_service.create_workspace(session, user_id, payload)
    return await _to_read_model(session, workspace)


@router.get("/me", response_model=WorkspaceRead)
async def read_my_workspace(
    workspace: Annotated[Workspace, Depends(get_current_workspace)],
    session: Annotated[AsyncSession, Depends(get_session)]
):
    return await _to_read_model(session, workspace)


@router.patch("/{workspace_id}/plan", response_model=WorkspaceRead)
async def update_plan(
    workspace_id: str,
    payload: WorkspacePlanUpdate,
    workspace: Annotated[Workspace, Depends(get_current_workspace)],
    session: Annotated[AsyncSession, Depends(get_session)]
):
    if workspace.id != workspace_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not your workspace")
    updated = await workspace_service.change_plan(session, workspace, payload.plan)
    return await _to_read_model(session, updated)

File: frontend/src/lib/api/workspaces.ts
import { apiFetch } from './client';
import type { PlanId } from '../../types';

export interface WorkspaceDTO {
  id: string;
  name: string;
  plan: PlanId;
  lead_quota: number;
  seat_quota: number;
  leads_used: number;
  seats_used: number;
  created_at: string;
}

export function getMyWorkspace(token: string | null) {
  return apiFetch<WorkspaceDTO>('/api/v1/workspaces/me', token);
}

export function createWorkspace(token: string | null, name: string, ownerEmail?: string, ownerName?: string) {
  return apiFetch<WorkspaceDTO>('/api/v1/workspaces', token, {
    method: 'POST',
    body: JSON.stringify({ name, owner_email: ownerEmail, owner_name: ownerName })
  });
}

export function updateWorkspacePlan(token: string | null, workspaceId: string, plan: PlanId) {
  return apiFetch<WorkspaceDTO>(`/api/v1/workspaces/${workspaceId}/plan`, token, {
    method: 'PATCH',
    body: JSON.stringify({ plan })
  });
}
